OSDOCS-2427: Update best practices guidance (#68477)
openshift/openshift-docs: modules/update-best-practices.adoc

// Module included in the following assemblies:
//
// * updating/preparing_for_updates/updating-cluster-prepare.adoc

:_mod-docs-content-type: PROCEDURE
[id="update-best-practices_{context}"]
= Best practices for cluster updates

{product-title} is designed to provide a robust update experience that allows clusters to update with minimal disruptions to workloads.
Updates will not begin unless the cluster is determined to be in an upgradeable state at the time of the update request.

While this design helps ensure that updates are successful as long as some key conditions are met, there are a number of actions you can take to increase your chances of a successful cluster update.

[discrete]
[id="recommended-versions_{context}"]
=== Choose versions recommended by the OpenShift Update Service

The OpenShift Update Service (OSUS) provides update recommendations based on cluster characteristics such as the cluster's subscribed channel, which are then saved by the Cluster Version Operator as either recommended or conditional updates.

> While it is possible to attempt an update to a version that is not recommended by OSUS, doing so significantly increases the risk of update failure or unintended consequences to the cluster after the update has finished.

Reviewer: Can we spin this into a positive-sounding (or at least less negative) statement? I think "significantly increases risk" is not even that much true. "update failure" is very strong term, but actually very rare, most upgrade issues are annoyances. I'd say recommended paths are near-zero risk and non-recommended is low risk, we just don't have enough data to actually claim near-zero risk (even if it likely is safe - we do a *lot* of testing, and we do not have enough  resources to test every possible path).

So I would maybe just say that following recommended paths protects the users from encountering known issues or unintended consequences, or something like that.

Author: I'll replace my wording with yours, sure thing.

Choose only update targets that are recommended by OSUS to ensure a successful update.

[discrete]
[id="critical-alerts_{context}"]
=== Address all critical alerts on the cluster

Critical alerts must always be addressed as soon as possible, but it is especially important to address these alerts and resolve any problems before initiating a cluster update.
Failing to address critical alerts before beginning an update can cause a loss of data or a major failure of cluster services.

> You should also periodically review Warning and Info alerts on the cluster to address any potentially problematic conditions before initiating an update.

Reviewer 2: I do not think we should ask customers to review info alerts , thats a lot of work without much value for update success. Just asking to review critical alerts is good enough.

Author: Sounds good, I'll remove this sentence.

[discrete]
[id="cluster-upgradeable_{context}"]
=== Ensure that the cluster is in an Upgradable state

When one or more Operators have not reported their `Upgradeable` condition as `true` for more than an hour, the `ClusterNotUpgradeable` warning alert is triggered in the cluster.

> In most cases patch updates are not blocked by this alert, but you cannot perform a minor version update until this alert is resolved and all Operators report `Upgradeable` as `true`.

Reviewer 3: nit: `true` -> `True`  ([API][1]).

[1]: https://github.com/openshift/api/blob/03f577bfbd59e4058724de268e5669db7ee3865b/config/v1/types_cluster_operator.go#L110

[discrete]
[id="nodes-ready_{context}"]

> === Ensure that all nodes are available

Reviewer: I think "all nodes are available" is not exactly the condition that we want. The point is that the cluster should not be running with little to no spare capacity. If the cluster is "full", then taking down a node to upgrade it can be either slow  or even impossible. So I would describe this section as "make sure the cluster has enough spare capacity so that the upgrade process can actually take some nodes offline to update them. This has two flavors, (you already mention one of them):
1. There is spare capacity on compute nodes, so that when taking a node offline temporarily, its workloads can be placed on other nodes comfortably
2. Worker pools do not have enough unavailable notes to prevent taking more nodes offline temporarily.

Reviewer 2: We have noticed that many times when a node is down MCO gets stuck. So I am supportive of the idea. I also agree that we should have spare nodes which would give some breathing room to pods.

// Completely guessing the explanation in this section just to have something to start with when this is reviewed by an SME.
Nodes should not be down when beginning an update.
Nodes that are not running and available may limit a cluster's ability to perform an update with minimal disruption to cluster workloads.

Depending the on the configured value of the cluster's `maxUnavailable` spec, an unavailable node can also prevent itself and other nodes from having machine configuration changes applied during a cluster update.

[discrete]
[id="pod-disruption-budget_{context}"]
=== Ensure that the cluster's PodDisruptionBudget is properly configured

The `PodDisruptionBudget` object allows you to define the minimum number or percentage of pod replicas that must be available at any given time.

> This configuration allows workloads to be protected from disruptions during maintenance tasks such as cluster updates.

Reviewer: I think we can afford to be a bit more specific:

1. Make sure that for HA workloads, there are replicas that can be temporarily taken offline without PDB configuration prohibiting that
2. Make  sure that non-HA workloads are either not protected by PDBs, or that there is an alternative reliable mechanism that allows draining these workloads eventually (such as periodic restart)

// Module included in the following assemblies:
//
// * updating/preparing_for_updates/updating-cluster-prepare.adoc

> :_mod-docs-content-type: PROCEDURE

Reviewer 4: ```suggestion
:_mod-docs-content-type: REFERENCE
```

Author: Good catch, I copied and pasted this metadata stuff from another module

[id="update-best-practices_{context}"]
= Best practices for cluster updates

> {product-title} is designed to provide a robust update experience that allows clusters to update with minimal disruptions to workloads.

Reviewer 4: ```suggestion
{product-title} provides a robust update experience that minimizes workload disruptions.
```
Reworded to avoid passive voice and the use of "allow" per ISG:
"Avoid stating that inanimate objects grant abilities to people, as in “the product allows you to…”. Whenever possible, use a direct, user-focused alternative such as “you can use the product to…” or “with this product, users can…”.

> Updates will not begin unless the cluster is determined to be in an upgradeable state at the time of the update request.

Reviewer 4: ```suggestion
Updates will not begin unless the cluster is in an upgradeable state at the time of the update request.
```
Passive voice

> This design ensures that some key conditions are met before initiating an update, but there are a number of actions you can take to increase your chances of a successful cluster update.

Reviewer 4: ```suggestion
This design enforces some key conditions before initiating an update, but there are a number of actions you can take to ensure a successful cluster update.
```
Reworded passive voice and added stronger tone. "increase your chances" sounds like the user is gambling

Author: The first part of your suggestion sounds good, but I'm actively avoiding using terms like "ensuring" to avoid making promises to the user about the success of an update. The spirit of this doc is to acknowledge factors that can affect the success of an update and inform users how they can mitigate risk factors, it's not to assuage users that their updates will be universally successful.

IMO while we wouldn't want to explicitly state that updates are a gamble, that's sort of the nature of major tasks that introduce change and instability to the cluster, such as an update, and we don't want to misrepresent that nature in a doc meant to give users best practices and advice.

[discrete]
[id="recommended-versions_{context}"]

> === Choose versions recommended by the OpenShift Update Service

Reviewer 5: I think it is better to let customers know how to get the OSUS recommended versions, like on web console or command line.

Author: Unless I'm mistaken - I think as long as users follow our update procedures then they'll always use recommended versions, unless users look up how to find conditional updates or use `--to-image`. But those two options take more work to find in the docs.

So I think we already show users how to get recommended versions because that's the default method that we document in almost all of our procedures.

Reviewer 5: um...you are right.

> === Choose versions recommended by the OpenShift Update Service

Author: I can change it to be a second level heading, sure. I made it a third level for stylistic reasons but it's not super important to keep it that way. The reason they don't show up in the navigation on the right, however is because of the `[discrete]` elements, such as the one on line 14. 

I put those there because I thought it would overwhelm the sidebar navigation, but I guess if there are only 5 of these sections then that's okay for now!

> === Choose versions recommended by the OpenShift Update Service

Reviewer 5: @skopacz1 I am sure you have updated the navigation level, but I can't see the change on preview page, could you help to confirm it?

> === Choose versions recommended by the OpenShift Update Service

Author: The automated preview page is not working correctly, sorry about that. I've manually uploaded the most recent preview with your feedback here: https://file.rdu.redhat.com/skopacz/OSDOCS-2427/updating/preparing_for_updates/updating-cluster-prepare.html#update-best-practices_updating-cluster-prepare

(you'll need to be on VPN to access this site)

> The OpenShift Update Service (OSUS) provides update recommendations based on cluster characteristics such as the cluster's subscribed channel, which are then saved by the Cluster Version Operator as either recommended or conditional updates.

Reviewer 4: ```suggestion
The OpenShift Update Service (OSUS) provides update recommendations based on cluster characteristics such as the cluster's subscribed channel. The Cluster Version Operator saves these recommendations as either recommended or conditional updates.
```
Rewording suggestion

The OpenShift Update Service (OSUS) provides update recommendations based on cluster characteristics such as the cluster's subscribed channel, which are then saved by the Cluster Version Operator as either recommended or conditional updates.
While it is possible to attempt an update to a version that is not recommended by OSUS, following a recommended update path protects users from encountering known issues or unintended consequences on the cluster.

Choose only update targets that are recommended by OSUS to ensure a successful update.

[discrete]
[id="critical-alerts_{context}"]

> === Address all critical alerts on the cluster

Reviewer 5: Where can customer find the critical alerts? and which alerts are critical? maybe a summarize here is helpful.

Reviewer 5: I think run `oc adm upgrade` can also find the critical alerts. /cc @shellyyang1989 for confirm.

Reviewer: Currently alerts are not surfaced in any `oc` command. OTA-1080 and OTA-1157 should eventually improve this state.

Failing to address critical alerts before beginning an update can cause problematic conditions for the cluster.

> In the *Administrator* perspective of the web console, critical alerts can be
> found in *Observe -> Alerting*.

Reviewer 4: ```suggestion
In the *Administrator* perspective of the web console, navigate to 
*Observe* -> *Alerting*.
```

Author: If this suggestion is meant to remove passive voice, how about I rewrite it to this instead?
```suggestion
In the *Administrator* perspective of the web console, you can find critical alerts in *Observe -> Alerting*.
```

This way it still mentions critical alerts, which is why I'm telling users how to go to the alerting section of the UI.

[discrete]
[id="cluster-upgradeable_{context}"]

> === Ensure that the cluster is in an Upgradable state

Reviewer 5: How do we know if a cluster is in an Upgradable state?

Author: We document the Upgradeable condition of cluster Operators here: [Understanding cluster Operator condition types](https://docs.openshift.com/container-platform/4.14/updating/understanding_updates/intro-to-updates.html#understanding_clusteroperator_conditiontypes_understanding-openshift-updates).

I can add a link to that page underneath this overall best practices section

Reviewer 5: That would be greatly help. Thanks!

> When one or more Operators have not reported their `Upgradeable` condition as `True` for more than an hour, the `ClusterNotUpgradeable` warning alert is triggered in the cluster.

Reviewer 4: ```suggestion
When one or more Operators have not reported their `Upgradeable` condition as `True` for more than an hour, the cluster triggers the `ClusterNotUpgradeable` warning alert.
```
passive voice

Author: Sounds good, let me just check whether saying that it's the cluster that triggers the warning is accurate enough.

Author: I've talked about this with an SME and the chain of events that cause this alert are complex and comprise multiple distinct components taking actions. I think I'll keep this passive, since the ISG states that the following two reasons, among others, are appropriate use cases for passive voice:
"Passive voice is acceptable when any of these conditions are true:
- The system performs the action.
- It is more appropriate to focus on the receiver of the action."

> In most cases patch updates are not blocked by this alert, but you cannot perform a minor version update until this alert is resolved and all Operators report `Upgradeable` as `True`.

Reviewer 4: ```suggestion
In most cases this alert does not block patch updates, but you cannot perform a minor version update until you resolve this alert and all Operators report `Upgradeable` as `True`.
```
passive voice

Author: Sounds good for the first part, but as to changing the second part to "until you resolve this alert": I'm honestly not familiar enough with OCP to suggest that this alert always requires user intervention to resolve. Maybe in some cases something weird happens to an Operator for an hour to trigger the Upgradeable=false condition, then it fixes itself, reports `Upgradeable` as `true`, and the alert goes away solely due to system actions? 

I think I'll keep it as "this alert is resolved" to keep that user vs system ambiguity, unless an SME can confirm that basically every time an Operator reports `Upgradeable` as `false`, it's an issue that requires user action to resolve.

CC @petr-muller and @wking

Author: Sounds good, I suppose the distinction is probably not relevant enough to leave that ambiguity in.

For more information about the `Upgradeable` condition, see "Understanding cluster Operator condition types" in the additional resources section.

[id="nodes-ready_{context}"]
== Ensure that enough spare nodes are available

// Completely guessing the explanation in this section just to have something to start with when this is reviewed by an SME.
A cluster should not be running with little to no spare node capacity, especially when initiating a cluster update.

> Nodes that are not running and available may limit a cluster's ability to perform an update with minimal disruption to cluster workloads.

Reviewer 4: I think it would be better to reword this in a positive voice, such as "You should have spare node capacity to ensure.." and "Having available nodes is important for cluster updates to complete with minimal disruption" or something to that effect. I'm not entirely sure what is meant by "running and available" though.

Author: I'll admit this sounds negative but I think it's useful to talk about the consequences this can have on update success, especially since there's already a more positive sounding, actionable piece of advice at the end of the section here on line 49. I feel like it would be a bit redundant to also have similar phrasing here. That's just my opinion tho, let me know if you disagree.

> Depending on the configured value of the cluster's `maxUnavailable` spec, an unavailable node can also prevent itself and other nodes from having machine configuration changes applied during a cluster update.

Reviewer 4: ```suggestion
Depending on the configured value of the cluster's `maxUnavailable` spec, the cluster may not be able to apply machine configuration changes to nodes if there is an unavailable node.
```
passive voice

Author: I can try and reword it in active voice, sure.

Additionally, if compute nodes do not have enough spare capacity, workloads might not be able to temporarily shift to another node while the first node is taken offline for an update.

Make sure that you have enough available nodes in each worker pool, as well as enough spare capacity on your compute nodes, to increase the chance of successful node updates.

[id="pod-disruption-budget_{context}"]
== Ensure that the cluster's PodDisruptionBudget is properly configured

> The `PodDisruptionBudget` object allows you to define the minimum number or percentage of pod replicas that must be available at any given time.

Reviewer 4: ```suggestion
You can use the `PodDisruptionBudget` object to define the minimum number or percentage of pod replicas that must be available at any given time.
```
Removed permissive language

> This configuration allows workloads to be protected from disruptions during maintenance tasks such as cluster updates.

Reviewer 4: ```suggestion
This configuration protects workloads from disruptions during maintenance tasks such as cluster updates.
```
